scritchPencilRaster: Fix MIDP transformations

AShiningRay opened the pull request:
Previously those were broken due to certain transforms not being processed at all, or under the assumption they'd reach the same result as another (was seemingly the case for ROT_180 and TRANS_MIRROR_ROT180). Furthermore, after those transforms were processed, even if the matrix transforms themselves were correct, the end result would still be incorrect on screen due to the source/destination positions and sizes not being updated to reflect the transformed data.

This commit fixes that, all transforms seem to be done properly now, in all cases known, passing the internal image tests as well as games such as CastleVania: Dawn of Sorrow (K800i version) and Sonic 2 Crash (also for K800i), both of which, together, use all the available transforms with a myriad of clipping and image width/heights.

Do note that Sonic 2 Crash will not yet boot at the current tree state due to SquirrelJME missing Nokia DirectGraphics' implementations as well as fillArc and fillRoundRect, those will be coming soon.

```
You grant Stephanie Gawroriski an irrevocable license that:

 1. That you own the contributing work.
 2. Grants a patent license, as per the GNU GPLv3.
 3. Granting Stephanie Gawroriski permission to redistribute, sell, lease,
    modify, transform, translate, and relicense the specified works. This
    is to simplify the licensing of the project and permit it to be
    consistent. Your contribution may be commercially licensed to other
    parties supporting the project through this means.
 4. If employed by a company, you have a right by that company to provide
    contributions to this project.
 5. Have pledged to follow the Code of Conduct.
 6. Have read and understand the Ettiquite.
```

 * I accept the contributor agreement.

 * I accept that SquirrelJME uses Fossil for its source control
   management and I do understand that my pull request will be merged
   via patch in Fossil instead of via GitHub.

AShiningRay committed:
scritchPencilRaster: Fix MIDP transformations

Previously those were broken due to certain transforms not being
processed at all, or under the assumption they'd reach the same
result as another (was seemingly the case for ROT_180 and
TRANS_MIRROR_ROT180). Furthermore, after those transforms were
processed, even if the matrix transforms themselves were correct,
the end result would still be incorrect on screen due to the
source/destination positions and sizes not being updated to reflect
the transformed data.

This commit fixes that, all transforms seem to be done properly now,
in all cases known, passing the internal image tests as well as games
such as CastleVania: Dawn of Sorrow (K800i version) and Sonic 2 Crash
(also for K800i), both of which, together, use all the available
transforms with a myriad of clipping and image width/heights.

Do note that Sonic 2 Crash will not yet boot at the current tree
state due to SquirrelJME missing Nokia DirectGraphics'
implementations as well as fillArc and fillRoundRect, those will
be coming soon.

XerTheSquirrel commented:
This is being tracked as `wip-ashiningray-transformfixes`.

XerTheSquirrel commented:
In this current state, this will have to be rejected. `sjme_scritchpen_core_drawXRGB32Region()` should not have the additional transformation changes. The transformation handling is less readable and less concise as well.

AShiningRay committed:
scritchPencilRaster: More improvements to transforms

This one is mostly janitorial tasks, like moving the coordinate
adjustments from drawXRGB32Region() into a separate applyCoordinateAdj()
function in pencilRasterMisc following what is done in applyRotateScale(),
bringing back the xform variable to keep things more in line with
the original logic, and making some cleanups to RotateScale, like
removing the "result" matrix which didn't serve much of a purpose
since we can just access outMatrix directly.

XerTheSquirrel commented:
This is much better, thank you! Will be integrating this at some point.

XerTheSquirrel commented:
Merged, thank you!
